Feat/dapp staking v3 phase2

Dinonard opened the pull request:
**Pull Request Summary**

Added new extrinsic calls:
* `unlock`
* `claim_unlocked`
* `relock_unlocking`

Refactored existing code to match `master` practices.
Partial adoption for the latest uplift (some TODOs remain).
Extended custom types with new functions, enums, traits and structs.

Lots of tests added, they make the bulk of the PR

shaunxw commented on pallets/dapp-staking-v3/src/types.rs:
```diff
@@ -170,108 +337,245 @@ where
             .map_or(EraNumber::zero(), |locked| locked.era)
     }
 
+    // TODO: can active_period be provided somehow different instead of using a parameter? It should be 'free' to read the data from storage since it will be whitelisted.
```
Reading from storage would make it a bit difficult to do unit tests.

Dinonard replied:
I was thinking of adding another type with a `Get<...>` but that would make it even more complex 🙈 . Will keep it like this.

shaunxw commented on pallets/dapp-staking-v3/src/types.rs:
```diff
@@ -170,108 +337,245 @@ where
+    pub fn claim_unlocked(&mut self, current_block_number: BlockNumber) -> Balance {
+        let mut total_unlocked = Balance::zero();
+
+        // Probably faster approach than partitioning, summing & collecting.
```
All good with the current impl, tho personally I would prefer to sacrifice a bit of performance for readability.

Dinonard replied:
Totally agree!

First I wrote it using approach with `iters`, but it was just...unreadable 😓

shaunxw replied:
`.partition` might help? Like:

```rust
pub fn claim_unlock(...) -> Balance {
    let (unlocked, unlocking): (Vec<_>, Vec<_>) = self.unlocking.iter().partition(|chunk| chunk.unlock_block <= current_block_number);
    // assign new unlocking
    // fold `unlocked`
}
```

Dinonard replied:
I did it like that, but after Rust formatter broke it down into many lines, it looked messy (to me). So I'd actually prefer to keep it like this.

shaunxw replied:
It should be fine with good test coverage.

Actually, for me the formatted code makes it easier to see the `partition` method and helps with readability.

shaunxw reviewed (commented):
Looks good in general. Some minor comments. Love the test coverage!

Dinonard commented on pallets/dapp-staking-v3/src/lib.rs:
```diff
@@ -89,6 +95,7 @@ pub mod pallet {
         type MaxNumberOfContracts: Get<DAppId>;
 
         /// Maximum number of locked chunks that can exist per account at a time.
+        // TODO: should this just be hardcoded to 2? Nothing else makes sense really - current era and next era are required.
```
I was thinking of clearing this later, just in case. I might be overlooking something and don't want to come back & return it later 😅 .

Dinonard commented on pallets/dapp-staking-v3/src/types.rs:
```diff
@@ -170,108 +337,245 @@ where
+    /// Total locked amount by the user.
+    /// Includes both active locked amount & unlocking amount.
+    pub fn total_locked_amount(&self) -> Balance {
+        self.active_locked_amount() + self.unlocking_amount()
```
Definitely, missed this one.

Dinonard commented on pallets/dapp-staking-v3/src/types.rs:
```diff
@@ -44,6 +45,162 @@ pub type PeriodNumber = u32;
 pub type DAppId = u16;
 
+/// Trait for types that can be used as a pair of amount & era.
+pub trait AmountEraPair: MaxEncodedLen + Default + Copy {
```
I'm not sure - yet. They are the same now, and I think they will remain like that.

I'll add a TODO to clean it up later if they remain the same.

Dinonard commented on pallets/dapp-staking-v3/src/types.rs:
```diff
@@ -44,6 +45,162 @@ pub type PeriodNumber = u32;
+///
+#[derive(Encode, Decode, MaxEncodedLen, Clone, Debug, PartialEq, Eq, TypeInfo)]
+#[scale_info(skip_type_params(ML))]
+// TODO: should I use `EncodeLike`?
```
I misunderstood this, will delete the comment.

Dinonard commented on pallets/dapp-staking-v3/src/types.rs:
```diff
@@ -44,6 +45,162 @@ pub type PeriodNumber = u32;
+            inner[index].saturating_reduce(amount);
+            index
+        } else {
+            let mut chunk = inner[index];
+            chunk.saturating_reduce(amount);
+            chunk.set_era(era);
+
+            inner.insert(index + 1, chunk);
```
Added some comments, let me know if if helps.

I also talked with Ash, and I said I should probably add some scenarios explaining what's happening in the rustdoc.

shaunxw replied:
Yeah adding some scenarios would help a lot for code review and maintenance.

Dinonard commented on pallets/dapp-staking-v3/src/types.rs:
```diff
@@ -44,6 +45,162 @@ pub type PeriodNumber = u32;
+        let mut i = relevant_chunk_index;
+        while i < inner.len() - 1 {
+            if inner[i].get_amount().is_zero() && inner[i + 1].get_amount().is_zero() {
+                inner.remove(i + 1);
```
Thanks, I even found a minor bug now 😂 


For the test scenarios - it's not possible to get more than two consecutive zero chunks. That would imply that 2 such chunks already existed in the first place.

shaunxw replied:
Haha, the updated impl looks good!

Dinonard commented:
@shaunxw I added more comments, and changed in-place vector modification to use `retain` instead.

Please take another look, and approve if all fine! 🙏

shaunxw reviewed (approved):
Looks good!

ashutoshvarma reviewed (approved):
Look good!
